timesheet_automatic_reminder/models/timesheet_reminder.py

```diff
@@ -0,0 +1,333 @@
+                    _el = end_leave
+                    _sl = datetime.combine(check_date,
+                                           datetime.strptime('00:00:00', TIME_FORMAT).time())\
+                        .replace(tzinfo=end_leave.tzinfo)
+                else:
+                    employee_leave_dict[employee.id][check_date.strftime(DATE_FORMAT)] += 9
+                    count += 1
+                    check_date += one_day
+                    continue
+                work_on = datetime.combine(check_date, work_start).replace(tzinfo=_sl.tzinfo)
+                work_off = datetime.combine(check_date, work_end).replace(tzinfo=_el.tzinfo)
+                leave_on = _sl if _sl > work_on else work_on
+                leave_off = _el if _el < work_off else work_off
+                leave_hours = leave_off.hour - leave_on.hour
+                employee_leave_dict[employee.id][check_date.strftime(DATE_FORMAT)] += \
+                    leave_hours if leave_hours > 0 else 0
+            count += 1
+            check_date += one_day
+        return employee_leave_dict
+
+    @api.model
+    def _get_employee_leaves(self, yesterday, count_days, work_start,
+                             work_end, employee_records=None):
+        """
+        Get the leave time about every employee
+        :param start_datetime:
+        :param end_datetime:
+        :param employee_records:
+        :return:
+        """
+        end_datetime = yesterday.strftime(DATETIME_FORMAT)
+        count_date = [yesterday.date()]
+        while len(count_date) < count_days:
+            s = (count_date[-1] - timedelta(days=1))
+            count_date.append(s)
+        start_datetime = count_date[-1].strftime(DATETIME_FORMAT)
+        holiday_obj = self.env['hr.holidays']
+        if not employee_records:
+            employee_records = self.env['hr.employee'].search([])
+        count_date_str = [r.strftime(DATE_FORMAT) for r in count_date]
+        employee_leave_dict = defaultdict(dict)
+        for employee in employee_records:
+            employee_leave_dict[employee.id] = dict.fromkeys(count_date_str, 0)
+        leave_records = holiday_obj.search([
+            ('employee_id', 'in', employee_records.ids),
+            ('type', '=', 'remove'),
+            ('state', 'not in', ['cancel', 'refuse']),
+            '|', '&',
+            ('date_from', '>=', start_datetime),
+            ('date_from', '<=', end_datetime),
+            '&',
+            ('date_to', '>=', start_datetime),
+            ('date_to', '<=', end_datetime)
+        ], order='date_from')
+        for leave in leave_records:
+            employee_leave_dict = self._get_leave_to_days(leave, work_start,
+                                                          work_end, employee_leave_dict)
+        return employee_leave_dict
+
+    @api.model
+    def _get_employee_tms(self, end_date, count_days, employee_records=False):
```

Comment:
@dongshukai is not about necessary or not, is about:
 
1. I need you be familiar with this mechanism.
2. Is easier to read to understand.
3. Is one step more on the paradigm of computation.
4. Odoo ORM will trigger computation if one of fields that depends on change, so is way easier to manage you'll have always the proper value.